File: LeetCode/0045. Jump Game II/solution.cpp
/**
 * 106 / 106 test cases passed.
 * Runtime: 12 ms
 * Memory Usage: 16.1 MB 
 */
class Solution {
public:
    int jump(vector<int>& nums) {
        int boundary = 0;
        int end = 0;
        int ans = 0;
        for (int i = 0; i < nums.size() - 1; i ++ ) {
            boundary = max(boundary, i + nums[i]);
            if (i == end) {
                end = boundary;
                ans ++;
            }
        }
        return ans;
    }
};

File: LeetCode/0045. Jump Game II/solution.py
"""
106 / 106 test cases passed.
Runtime: 48 ms
Memory Usage: 15.5 MB
"""
class Solution:
    def jump(self, nums: List[int]) -> int:
        farthest, ans, end = 0, 0, 0
        for i, num in enumerate(nums[:-1]):
            farthest = max(farthest, i + num)
            if i == end:
                end = farthest
                ans += 1
        return ans
